### PR TITLE
fix: json qb array expression

### PR DESCRIPTION
**Pull request by piyushsingariya**
## Pull Request

---

### 📄 Summary
fixes JSON QB array expressions

Closes signoz/engineering-pod#3813

---

### ✅ Change Type
_Select all that apply_

- [ ] ✨ Feature
- [x] 🐛 Bug fix
- [ ] ♻️ Refactor
- [ ] 🛠️ Infra / Tooling
- [ ] 🧪 Test-only

---

- Tests added/updated:
- Manual verification:
- Edge cases covered:

---

| Field | Value |
|------|-------|
| Deployment Type | Cloud / OSS / Enterprise |
| Change Type | Feature / Bug Fix / Maintenance |
| Description | User-facing summary |

---

### 📋 Checklist
- [ ] Tests added or explicitly not required
- [ ] Manually tested
- [ ] Breaking changes documented
- [ ] Backward compatibility considered

---

## 👀 Notes for Reviewers

---

> [!NOTE]
> **Medium Risk**
> Touches query generation for JSON filters (including promoted vs. non-promoted fallback) and changes operator semantics for `IN`/`BETWEEN`, which could subtly alter results/performance for log searches.
> 
> **Overview**
> Fixes incorrect SQL generation for JSON query-builder filters on array paths, especially when mixing promoted `body_json_promoted` data with `body_json` fallback.
> 
> The field/condition builders now (1) fall back from promoted arrays only when the promoted array is non-empty (instead of `coalesce`), (2) generate valid `arrayExists` lambdas, and (3) support `IN`/`NOT IN` and new `BETWEEN`/`NOT BETWEEN` operators for JSON conditions using `sb.In`/`sb.Between`.
> 
> Metadata/type inference is updated to better treat array-path promotion and array-valued operators (adds `FilterOperator.IsArrayOperator`, array scalar→array datatype mapping, and improved `InferDataType`), with tests updated/added to cover `IN`/`BETWEEN`, promoted-array fallback, and revised expected SQL.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit d1146b886a30477051310fe098843f4d4462f61d. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: pkg/telemetrylogs/json_stmt_builder_test.go [hunk 1]
```diff
@@ -316,7 +316,7 @@ func TestStatementBuilderListQueryBody(t *testing.T) {
...
...tes_bool, resources_string, scope_string FROM signoz_logs.distributed_logs_v2 WHERE resource_fingerprint GLOBAL IN (SELECT fingerprint FROM __resource_filter) AND ((arrayExists(`body_json.education`-> (arrayExists(x -> LOWER(toString(x)) LIKE LOWER(?), dynamicElement(`body_json.education`.`parameters`, 'Array(Nullable(Float64))')) OR arrayExists(x -> (toFloat64(x) = ?), dynamicElement(`body_json.education`.`parameters`, 'Array(Nullable(Float64))'))), dynamicElement(body_json.`education`, 'Array(JSON(max_dynamic_types=16, max_dynamic_paths=0))'))) OR (arrayExists(`body_json.education`-> (arrayExists(x -> LOWER(toString(x)) LIKE LOWER(?), arrayMap(x->dynamicElement(x, 'Float64'), arrayFilter(x->(dynamicType(x) = 'Float64'), dynamicElement(`body_json.education`.`parameters`, 'Array(Dynamic)')))) OR arrayExists(x -> (toFloat64(x) = ?), arrayMap(x->dynamicElement(x, 'Float64'), arrayFilter(x->(dynamicType(x) = 'Float64'), dynamicElement(`body_json.education`.`parameters`, 'Array(Dynamic)'))))), dynamicElement(body_json.`education`, 'Array(JSON(max_dynamic_types=16, max_dynamic_paths=0))')))) AND timestamp >= ? AND ts_bucket_start >= ? AND timestamp < ? AND ts_bucket_start <= ? LIMIT ?",
```
Comment (nityanandagohain) [inline pkg/telemetrylogs/json_stmt_builder_test.go]: I guess this seems wrong 
```
x -> (toFloat64(x) = ?)
```

**Review by nityanandagohain — COMMENTED** [review]
have a look at the arrayExists, it feels incorrect

File: pkg/telemetrylogs/json_string.go [hunk 2]
```diff
@@ -60,11 +67,6 @@ func InferDataType(value any, operator qbtypes.FilterOperator, key *telemetrytyp
 		valueType = telemetrytypes.FieldDataTypeBool
 	}
 
-	// check if it is array
-	if strings.HasSuffix(key.Name, "[*]") || strings.HasSuffix(key.Name, "[]") {
-		valueType = telemetrytypes.FieldDataType{String: valuer.NewString(fmt.Sprintf("[]%s", valueType.StringValue()))}
-	}
-
```
Comment (srikanthccv) [inline pkg/telemetrylogs/json_string.go:67]: This is a backward incompatible change. I don't see anything mentioned about it in the PR description. By removing this change, we are saying that the key name ending with `[*]` or `[]` won't decide if something is an array, rather it's the value type, which is incorrect. The tests are not failing because the where clause visitor sends the `list` of `any` for the value type, but that indirect implementation detail should not be relied upon here. Where clause visitor could change it's implementation but the `GetBodyJSONKey` should still work as expected.

What should following return?
```
InferDataType(1, qbtypes.FilterOperatorIn, {'name': 'body.user_names[*]'})
```

It return `Int64`, which is wrong. the `InferDataType` used for correct key building expression in the current logs query builder.

File: pkg/telemetrylogs/json_string.go [hunk 3]
```diff
@@ -41,31 +41,50 @@ func parseStrValue(valueStr string, operator qbtypes.FilterOperator) (telemetryt
...
+			}
+
+			arrayType := telemetrytypes.ScalerFieldTypeToArrayFieldType[scalerType]
+			switch {
+			// decide on the field data type based on the key
+			case key.FieldDataType.IsArray():
+				return arrayType, v
+			default:
+				// TODO(Piyush): backward compatibility for the old String based JSON QB queries
+				if strings.HasSuffix(key.Name, telemetrytypes.ArrayAnyIndexSuffix) {
```
[the remark this answers is not in the document]
Comment (piyushsingariya) [reply pkg/telemetrylogs/json_string.go:73]: This can be ignored, OLD JSON QB only works with `[*]`

File: pkg/telemetrylogs/json_string.go [hunk 4]
```diff
@@ -41,31 +41,50 @@ func parseStrValue(valueStr string, operator qbtypes.FilterOperator) (telemetryt
...
+		valueType := telemetrytypes.FieldDataTypeUnspecified
+		switch v := value.(type) {
+		case []any:
+			// take the first element and infer the type
+			var scalerType telemetrytypes.FieldDataType
+			if len(v) > 0 {
+				scalerType, _ = closure(v[0], key)
+			}
+
+			arrayType := telemetrytypes.ScalerFieldTypeToArrayFieldType[scalerType]
```
Comment (tushar-signoz) [inline pkg/telemetrylogs/json_string.go:66]: This would break if the value is an array of array of scalar. Lets handle this case to future proof this method

**Review by tushar-signoz — APPROVED** [review]
LGTM

**Review by srikanthccv — APPROVED** [review]
